molvis/molvisbase/src/datastructures/molecularstructure.cpp

```diff
@@ -0,0 +1,229 @@
+
+const std::vector<Chain>& MolecularStructure::getChains() const { return chains_; }
+
+std::optional<Residue> MolecularStructure::getResidue(size_t residueId, size_t chainId) const {
+    auto it = util::find_if(residues_,
+                            [&](auto& r) { return (r.id == residueId) && (r.chainId == chainId); });
+    if (it != residues_.end()) {
+        return *it;
+    } else {
+        return std::nullopt;
+    }
+}
+
+std::optional<Chain> MolecularStructure::getChain(size_t chainId) const {
+    auto it = util::find_if(chains_, [id = chainId](auto& r) { return r.id == id; });
+    if (it != chains_.end()) {
+        return *it;
+    } else {
+        return std::nullopt;
+    }
+}
+
+std::optional<size_t> MolecularStructure::getAtomIndex(const std::string& fullAtomName,
+                                                       size_t residueId, size_t chainId) const {
+    auto residue = getResidue(residueId, chainId);
+    if (residue) {
+        for (auto i : residue->atoms) {
+            if ((atoms_.fullNames[i] == fullAtomName) && (atoms_.residueIds[i] == residueId) &&
+                (atoms_.chainIds[i] == chainId)) {
+                return i;
+            }
+        }
+    } else {
+        for (size_t i = 0; i < atoms_.fullNames.size(); ++i) {
+            if ((atoms_.fullNames[i] == fullAtomName) && (atoms_.residueIds[i] == residueId) &&
+                (atoms_.chainIds[i] == chainId)) {
+                return i;
+            }
+        }
+    }
+    return std::nullopt;
+}
+
+void MolecularStructure::updateStructure() {
+    atoms_.residueIndices.clear();
+
+    for (size_t i = 0; i < atoms_.size(); ++i) {
+        const auto resId = atoms_.residueIds[i];
+        const auto chainId = atoms_.chainIds[i];
+        auto it = util::find_if(residues_, [id = resId](const Residue& r) { return r.id == id; });
+        if (it == residues_.end()) {
+            throw Exception(fmt::format("Invalid residue ID '{}'", resId), IVW_CONTEXT);
+        }
+        IVW_ASSERT(chainId == it->chainId, "Chain ID mismatch");
+
+        it->atoms.insert(i);
+        const auto residueIndex = std::distance(residues_.begin(), it);
+        atoms_.residueIndices.push_back(static_cast<int>(residueIndex));
+
+        auto chainIt =
+            util::find_if(chains_, [id = chainId](const Chain& c) { return c.id == id; });
+        if (chainIt == chains_.end()) {
+            throw Exception(fmt::format("Invalid chain ID '{}'", chainId));
+        }
+        chainIt->residues.insert(residueIndex);
+    }
+}
+
+void MolecularStructure::computeCovalentBonds() {
+    bonds_.clear();
+    if (atoms_.empty()) return;
+
+    const double maxCovalentBondLength = 4.0;
+    const dvec3 cellSize{maxCovalentBondLength};
+
+    // create uniform grid enclosing all atoms +- 1.0
+    dvec3 min{atoms_.positions.front()};
+    dvec3 max{atoms_.positions.front()};
+    for (size_t i = 0; i < atoms_.size(); ++i) {
```

Comment:
std::minmax_element

Reply:
would not work, since minmax_element refers to the min and max element. But here we are looking for the component-wise min/max.